### PR TITLE
Added models for user profiles

### LINKED ISSUE #18: create model & API for mm user profile
may need to talk to Erdin to understand scope & types of fields

foreign key to the user

for example:

GET /profiles/user_id
POST /profiles/
PATCH /profiles/user_id

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Connected to #18 

#### What's this PR do?
Adds models for the user profile. This PR does not include any REST API work

#### Where should the reviewer start?
profiles/models.py

#### What GIF best describes this PR or how it makes you feel?
https://i.imgur.com/CaG0KPC.gifv

### CODE REVIEW

File: profiles/api.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+"""
+from rest_framework import serializers, viewsets
+from profiles.models import Profile
+
+
+class ProfileSerializer(serializers.ModelSerializer):
+    """Serializer for Program objects"""
+    class Meta:  # pylint: disable=missing-docstring
+        model = Profile
+        fields = ('id',)
```
Comment (giocalitri) [inline profiles/api.py]: I think this needs to go in a `serializers.py` file

File: profiles/api.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class ProfileViewSet(viewsets.ReadOnlyModelViewSet):
+    """API for the Program collection"""
+    queryset = Profile.objects.all()
+    serializer_class = ProfileSerializer
```
Comment (giocalitri) [inline profiles/api.py]: is this the REST view? if so it should be in a `views.py` file

I understand what you meant to do, but in the past the `api.py` module contained the api functions used across the app and not the REST APIs, so I would avoid this approach.

Another comment is that we might want to version the rest APIs using directories path like `/api/v0/...`

File: profiles/models.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from django.contrib.auth.models import User
+from django.db import models
+
+
+class Profile(models.Model):
+    """
+    A user profile
+    """
+    user = models.ForeignKey(User)
+    edx_username = models.TextField()
```
Comment (giocalitri) [inline profiles/models.py]: why is this here? can it be even possible to have a different username from edx?

File: profiles/api.py (hunk 2 above)
Comment (noisecapella) [inline profiles/api.py]: Ok, we should also change `courses/api.py` to `courses/serializers.py` and `courses/views.py` to be consistent

**Comment by bdero** [comment]
This should probably be rebased on top of https://github.com/mitodl/micromasters/pull/39 so that the endpoints are verifiable on the PR app.

File: profiles/models.py (hunk 3 above)
Comment (noisecapella) [inline profiles/models.py]: It seems like we can't rely on the two usernames always being identical, but you probably know better than me. When the user logs in via OAuth is the User object created automatically?

Comment (justinabrahms) [inline profiles/models.py]: Yes. If that user is already logged in, they get an additional account associated to them.

This username will be queryable via python social auth.

File: profiles/models.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    PUBLIC_TO_MM = 'public_to_mm'
+    PUBLIC = 'public'
+    ACCOUNT_PRIVACY_CHOICES = (
+        (PRIVATE, 'Private'),
+        (PUBLIC_TO_MM, 'Public to logged in users'),
+        (PUBLIC, 'Public to everyone'),
+    )
+
+    user = models.ForeignKey(User)
+    account_privacy = models.TextField(
```
Comment (giocalitri) [inline profiles/models.py]: `TextField` sounds a bit excessive: why not `CharField`?

File: profiles/models.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        default=PRIVATE,
+        choices=ACCOUNT_PRIVACY_CHOICES,
+    )
+    email_optin = models.BooleanField(default=False)
+    employer = models.TextField(blank=True, null=True)
+    job_title = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: also for `employer` and `job_title` maybe `TextField` is too much

File: profiles/models.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    state_or_territory = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

File: profiles/models.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    name = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

File: profiles/models.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    bio = models.TextField(blank=True, null=True)
+    country = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

File: profiles/models.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    has_profile_image = models.BooleanField()
+    profile_url_full = models.TextField(blank=True, null=True)
+    profile_url_large = models.TextField(blank=True, null=True)
+    profile_url_medium = models.TextField(blank=True, null=True)
+    profile_url_small = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField` for all urls

File: profiles/models.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    requires_parental_consent = models.NullBooleanField()
+    year_of_birth = models.IntegerField(blank=True, null=True)
+    level_of_education = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

File: profiles/models.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    goals = models.TextField(blank=True, null=True)
+    language_proficiencies = JSONField(blank=True, null=True)
+    gender = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

File: profiles/models.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    mailing_address = models.TextField(blank=True, null=True)
```
Comment (giocalitri) [inline profiles/models.py]: some here about `TextField`

**Comment by giocalitri** [comment]
I think some fields of the model should not be declared as `TextField`, but in general the PR is in a pretty good shape.

File: profiles/models.py (hunk 4 above)
Comment (noisecapella) [inline profiles/models.py]: TextField and CharField are identical under the hood for postgres databases

Comment (pwilkins) [inline profiles/models.py]: I didn't read the StackOverflow answer to mean that they are identical in Postgres, just that there was no performance difference.  For Django there are fewer functional differences than if we weren't using an ORM.

File: profiles/models.py (hunk 9 above)
Comment (noisecapella) [inline profiles/models.py]: I'm a bit reluctant to use UrlField or anything else here since these fields are copied from edX (at least initially) so they're not guaranteed to validate

File: profiles/models.py (hunk 4 above)
Comment (giocalitri) [inline profiles/models.py]: they are the same if you do not put a max_length and it does not make much sense to not doing it when we know the max size of the field, like in this case and many of the following.

anyway feel free to leave it in this way if you have a strong opinion about it.

File: profiles/models.py (hunk 9 above)
Comment (giocalitri) [inline profiles/models.py]: I agree about `UrlField`

File: profiles/models.py (hunk 12 above)
Comment (noisecapella) [inline profiles/models.py]: It doesn't look like there's any max length in the edX model where this information comes from, so I'm reluctant to have a max length here

**Comment by noisecapella** [comment]
I changed some of the fields to CharField with max_length=255, and gender and level_of_education have choices and max_length which matches the UserProfile model in edX

**Comment by giocalitri** [comment]
fine with me :+1:
